packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -0,0 +1,618 @@
+
+    // token -> tokenID
+    mapping (address => uint16)                public cachedTokenIDs;
+
+    event Transfers(uint batchID, InternalBridgeTransfer[] transfers);
+
+    event BridgeCallSuccess (address connector);
+    event BridgeCallFailed  (address connector, string reason);
+
+    event ConnectorTrusted  (address connector, bool trusted);
+
+    modifier onlyFromExchangeOwner()
+    {
+        require(msg.sender == exchangeOwner, "UNAUTHORIZED");
+        _;
+    }
+
+    constructor(
+        IExchangeV3      _exchange,
+        uint32           _accountID
+        )
+    {
+        exchange = _exchange;
+        accountID = _accountID;
+
+        depositContract = _exchange.getDepositContract();
+        exchangeOwner = _exchange.owner();
+
+        DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("Bridge", "1.0", address(this)));
+    }
+
+    function batchDeposit(
+        BridgeTransfer[] memory deposits
+        )
+        external
+        payable
+    {
+        // Needs to be possible to do all transfers in a single block
+        require(deposits.length <= MAX_NUM_TRANSACTIONS_IN_BLOCK, "MAX_DEPOSITS_EXCEEDED");
+
+        // Transfers to be done
+        InternalBridgeTransfer[] memory transfers = new InternalBridgeTransfer[](deposits.length);
+
+        // Worst case scenario all tokens are different
+        TokenData[] memory tokens = new TokenData[](deposits.length);
+        uint numDistinctTokens = 0;
+
+        // Run over all deposits summing up total amounts per token
+        address token = address(-1);
+        uint tokenIdx = 0;
+        for (uint i = 0; i < deposits.length; i++) {
+            if(token != deposits[i].token) {
+                token = deposits[i].token;
+                tokenIdx = 0;
+                while(tokenIdx < numDistinctTokens && tokens[tokenIdx].token != token) {
+                    tokenIdx++;
+                }
+                if (tokenIdx == numDistinctTokens) {
+                    tokens[tokenIdx].token = token;
+                    tokens[tokenIdx].tokenID = _getTokenID(token);
+                    numDistinctTokens++;
+                }
+            }
+            tokens[tokenIdx].amount = tokens[tokenIdx].amount.add(deposits[i].amount);
+            deposits[i].token = address(tokens[tokenIdx].tokenID);
+
+            transfers[i].owner = deposits[i].owner;
+            transfers[i].tokenID = tokens[tokenIdx].tokenID;
+            transfers[i].amount = deposits[i].amount;
+        }
+
+        // Do a normal deposit per token
+        for(uint i = 0; i < numDistinctTokens; i++) {
+            if (tokens[i].token == address(0)) {
+                require(tokens[i].amount == msg.value, "INVALID_ETH_DEPOSIT");
+            }
+            _deposit(msg.sender, tokens[i].token, uint96(tokens[i].amount));
+        }
+
+        // Store the transfers so they can be processed later
+        _storeTransfers(transfers);
+    }
+
+    function beforeBlockSubmission(
```

Comment:
this is the bridged call to L2 dapp, is it? I almost confused it with `batchDeposit`.

If `batchDeposit` and `beforeBlockSubmission` are two independed features, maybe we should extract `batchDeposit` to a dedicated file...

Reply:
> this is the bridged call to L2 dapp, is it? I almost confused it with batchDeposit.

To L1 dapps yes.

> If batchDeposit and beforeBlockSubmission are two independed features, maybe we should extract batchDeposit to a dedicated file

`batchDeposit` and `BridgeCall`s are quite related. `batchDeposit` is for L1 -> L2, `BridgeCall`s is for L2 -> L1, but often that also requires and L1 -> L2 operation, which is done using  `batchDeposit`. Both are indeed done in the Bridge contract, making it a two-way bridge, which I think makes sense. I think splitting them up could also work, but probably a bit less efficient.

Reply:
I am ok with not spling the code into multiple ones.